jupyterhub/objects.py

```diff
@@ -26,10 +27,22 @@ class Server(HasTraits):
     orm_server = Instance(orm.Server, allow_none=True)
 
     ip = Unicode()
+    connect_ip = Unicode()
     proto = Unicode('http')
     port = Integer()
     base_url = Unicode('/')
     cookie_name = Unicode('')
+    
+    @property
+    def _connect_ip(self):
+        """Property for connect_ip"""
+        if self.connect_ip:
+            return self.connect_ip
+        elif self.ip in {'', '0.0.0.0'}:
+            # if listening on all interfaces, default to hostname
```

Comment:
Nothing here will bind to a hostname. I'll add some comments that clarify that `connect_ip` is only ever used for connecting, never binding. Nothing about binding is changed in this PR and users can still specify whatever they want for the bind address. This is only for "If bound on all interfaces, tell clients to connect via hostname unless another ip is specified for connections."

The reasoning behind switching from 127.0.0.1 to hostname when the bind address is all-interfaces:

The default for this case so far has been 127.0.0.1, which works if everything is local and doesn't work when clients are remote. If the bind address is explicitly changed to 0.0.0.0, it is probable that there are remote clients (otherwise, why choose it?). The result up to now is that that setting `ip = 0.0.0.0` *always requires* setting a different config value for the connect address explicitly as well (e.g. `DockerSpawner.hub_ip_connect`), depending on the Spawner. This default case is changed to `hostname`, which will work for remote and local clients in sensible networking conditions, and can always be overridden with `JupyterHub.hub_connect_ip`, just like DockerSpawner and KubeSpawner have been doing themselves.